### PR TITLE
WIP: [ECHOT-83] add proposal operations 

### PR DESCRIPTION
**Pull request by GwinSalbor**
WIP: [ECHOT-83] add proposal operations

### CODE REVIEW

File: common/base_test.py [hunk 1]
```diff
@@ -498,22 +498,24 @@ def get_accounts_ids(self, account_name, account_count, database_api_identifier,
...
 
     def add_fee_to_operation(self, operation, database_api_identifier, fee_amount=None, fee_asset_id="1.3.0",
-                             debug_mode=False):
+                             proposal=False, debug_mode=False):
         try:
             if fee_amount is None:
-                fee = self.get_required_fee(operation, database_api_identifier, asset=fee_asset_id,
-                                            debug_mode=debug_mode)
+                fee = self.get_required_fee(operation, database_api_identifier, asset=fee_asset_id, proposal=proposal,
+                                            debug_mode=True)
```
Comment (yauheni-vasilyeu) [inline common/base_test.py]: debug_mode=debug_mode

File: common/echo_operation.py [hunk 2]
```diff
@@ -351,13 +354,12 @@ def get_asset_claim_fees_operation(self, echo, issuer, claim_amount=0, claim_ass
         return [operation_id, asset_claim_fees_props, signer]
 
     def get_proposal_create_operation(self, echo, fee_paying_account, expiration_time, proposed_ops,
-                                      review_period_seconds=None, fee_amount=0, fee_asset_id="1.3.0",
-                                      extensions=None, signer=None, debug_mode=False):
+                                      review_period_seconds=None, extensions=None, signer=None, debug_mode=False):
```
Comment (yauheni-vasilyeu) [inline common/echo_operation.py:357]: Write api-doc comment about, not using 'collect' for this operation

File: common/echo_operation.py [hunk 3]
```diff
@@ -367,10 +369,46 @@ def get_proposal_create_operation(self, echo, fee_paying_account, expiration_tim
...
+        del proposal_create_props["fee"]
         if signer is None:
             return [operation_id, proposal_create_props, fee_paying_account]
         return [operation_id, proposal_create_props, signer]
 
+
+    def get_proposal_update_operation(self, echo, fee_paying_account, proposal, active_approvals_to_add,
+                                      active_approvals_to_remove, key_approvals_to_add, key_approvals_to_remove,
+                                      fee_amount=0, fee_asset_id="1.3.0",
+                                      extensions=None, signer=None, debug_mode=False):
```
Comment (yauheni-vasilyeu) [inline common/echo_operation.py]: fix tabulation

File: common/echo_operation.py [hunk 4]
```diff
@@ -367,10 +369,46 @@ def get_proposal_create_operation(self, echo, fee_paying_account, expiration_tim
...
+        operation_id = echo.config.operation_ids.PROPOSAL_DELETE
+        proposal_delete_props = self.get_operation_json("proposal_delete_operation")
+        proposal_delete_props["fee"].update({"amount": fee_amount, "asset_id": fee_asset_id})
+        proposal_delete_props.update({"fee_paying_account": fee_paying_account,
+                                      "using_owner_authority": using_owner_authority, "proposal": proposal})
+        if debug_mode:
+            lcc.log_debug("Proposal delete operation: \n{}".format(json.dumps(proposal_delete_props, indent=4)))
+        if signer is None:
+            return [operation_id, proposal_delete_props, fee_paying_account]
+        return [operation_id, proposal_delete_props, signer]
```
Comment (yauheni-vasilyeu) [inline common/echo_operation.py:410]: blank line

File: common/echo_operation.py [hunk 5]
```diff
@@ -772,10 +810,15 @@ def broadcast(self, echo, list_operations, return_operations=False, expiration=N
...
-            tx.add_signer(self.get_signer(signer=operation[2]))
+            if type(operation[2]) is list:
+                for i, op in enumerate(operation[2]):
+                    tx.add_signer(self.get_signer(signer=operation[2][i]))
+            else:
+                tx.add_signer(self.get_signer(signer=operation[2]))
         if expiration:
             tx.expiration = expiration
         tx.sign()
+        lcc.log_info("{}".format(tx.transaction_object))
```
Comment (yauheni-vasilyeu) [inline common/echo_operation.py]: remove log_info

File: common/echo_operation.py [hunk 6]
```diff
@@ -367,10 +369,46 @@ def get_proposal_create_operation(self, echo, fee_paying_account, expiration_tim
             proposal_create_props.update({"review_period_seconds": review_period_seconds})
         if debug_mode:
             lcc.log_debug("Proposal create operation: \n{}".format(json.dumps(proposal_create_props, indent=4)))
+        del proposal_create_props["fee"]
```
Comment (yauheni-vasilyeu) [inline common/echo_operation.py]: delete this line
remove "fee" object for proposal_create operation in echo_operations.json

File: suites/Operations/ProposalOperations/ProposalCreate.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+class ProposalCreate(BaseTest):
+
+    def __init__(self):
+        super().__init__()
+        self.__database_api_identifier = None
+        self.__registration_api_identifier = None
+        self.echo_acc0 = None
+        self.echo_acc1 = None
+
+    def get_expiration_time(self, seconds):
```
Comment (yauheni-vasilyeu) [inline suites/Operations/ProposalOperations/ProposalCreate.py]: move this function to **base_test**
remove from all proposal tests (include get_proposed_transactions api method suite)

File: suites/Operations/ProposalOperations/ProposalCreate.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+            proposed_ops=collected_operation,
+            expiration_time=self.get_expiration_time(15),
+            review_period_seconds=10
+        )
+        broadcast_result = self.echo_ops.broadcast(echo=self.echo, list_operations=operation)
+
+        require_that(
+            "broadcast transaction complete successfully",
+            self.is_operation_completed(broadcast_result, 1), is_true()
+        )
```
Comment (yauheni-vasilyeu) [inline suites/Operations/ProposalOperations/ProposalCreate.py:64]: add quite=True

File: suites/Operations/ProposalOperations/ProposalCreate.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+                                        self.__database_api_identifier)
+        response = self.get_response(response_id)
+        lcc.log_info("Call 'get_proposed_transactions' with id='{}' parameter".format(self.echo_acc1))
+
+        proposed_transactions_ids = [propose_transaction["id"] for propose_transaction in response["result"]]
+        lcc.set_step("Check 'get_proposed_transactions' method result")
+        require_that(
+            "proposed transactions ids",
+            proposed_transactions_ids, has_item(proposal_id)
+        )
```
Comment (yauheni-vasilyeu) [inline suites/Operations/ProposalOperations/ProposalCreate.py:78]: quite=True

File: suites/Operations/ProposalOperations/ProposalCreate.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+from common.base_test import BaseTest
+
+SUITE = {
+    "description": "Operation 'proposal_create'"
+}
+
+
+@lcc.prop("main", "type")
+@lcc.tags("operations", "asset_transfer_operations", "proposal_create")
```
Comment (yauheni-vasilyeu) [inline suites/Operations/ProposalOperations/ProposalCreate.py]: not "asset_transfer_operations"

File: suites/Operations/ProposalOperations/ProposalCreate.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+    @lcc.test("Simple work of method 'proposal_create'")
+    def method_main_check(self):
+        transfer_operation = self.echo_ops.get_transfer_operation(echo=self.echo,
+                                                                  from_account_id=self.echo_acc1,
+                                                                  to_account_id=self.echo_acc0)
+        lcc.set_step("Collect transfer operation")
+        collected_operation = self.collect_operations(transfer_operation, self.__database_api_identifier)
+
+        lcc.set_step("Make proposal of new active committee member")
```
Comment (yauheni-vasilyeu) [inline suites/Operations/ProposalOperations/ProposalCreate.py]: change step text

File: suites/Operations/ProposalOperations/ProposalDelete.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+from common.base_test import BaseTest
+
+SUITE = {
+    "description": "Operation 'proposal_delete'"
+}
+
+
+@lcc.prop("main", "type")
+@lcc.tags("operations", "asset_transfer_operations", "proposal_delete")
```
Comment (yauheni-vasilyeu) [inline suites/Operations/ProposalOperations/ProposalDelete.py]: asset_transfer_operations => change

File: suites/Operations/ProposalOperations/ProposalDelete.py [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+class ProposalDelete(BaseTest):
+
+    def __init__(self):
+        super().__init__()
+        self.__database_api_identifier = None
+        self.__registration_api_identifier = None
+        self.echo_acc0 = None
+        self.echo_acc1 = None
+
+    def get_expiration_time(self, seconds):
```
Comment (yauheni-vasilyeu) [inline suites/Operations/ProposalOperations/ProposalDelete.py]: like in proposal_create suite

File: suites/Operations/ProposalOperations/ProposalDelete.py [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+    def setup_suite(self):
+        super().setup_suite()
+        self._connect_to_echopy_lib()
+        lcc.set_step("Setup for {}".format(self.__class__.__name__))
+        self.__database_api_identifier = self.get_identifier("database")
+        self.__registration_api_identifier = self.get_identifier("registration")
+        lcc.log_info(
+            "API identifiers are: database='{}', registration='{}'".format(self.__database_api_identifier,
+                                                                           self.__registration_api_identifier))
+        self.echo_acc0 = self.get_account_id(self.accounts[0], self.__database_api_identifier,
```
Comment (yauheni-vasilyeu) [inline suites/Operations/ProposalOperations/ProposalDelete.py]: remove accounts

File: suites/Operations/ProposalOperations/ProposalDelete.py [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+                                                                  amount=transfer_amount, to_account_id="1.2.6")
+        collected_operation = self.collect_operations(transfer_operation, self.__database_api_identifier)
+        lcc.log_info("Transfer operation collected")
+
+        lcc.set_step("Make proposal of transfer operation")
+        operation = self.echo_ops.get_proposal_create_operation(
+            echo=self.echo,
+            fee_paying_account="1.2.6",
+            proposed_ops=collected_operation,
+            expiration_time=self.get_expiration_time(1000),
```
Comment (yauheni-vasilyeu) [inline suites/Operations/ProposalOperations/ProposalDelete.py]: 60

File: suites/Operations/ProposalOperations/ProposalDelete.py [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+        self.__database_api_identifier = self.get_identifier("database")
+        self.__registration_api_identifier = self.get_identifier("registration")
+        lcc.log_info(
+            "API identifiers are: database='{}', registration='{}'".format(self.__database_api_identifier,
+                                                                           self.__registration_api_identifier))
+        self.echo_acc0 = self.get_account_id(self.accounts[0], self.__database_api_identifier,
+                                             self.__registration_api_identifier)
+        self.echo_acc1 = self.get_account_id(self.accounts[1], self.__database_api_identifier,
+                                             self.__registration_api_identifier)
+        lcc.log_info("Echo accounts are: #1='{}', #2='{}'".format(self.echo_acc0, self.echo_acc1))
```
Comment (yauheni-vasilyeu) [inline suites/Operations/ProposalOperations/ProposalDelete.py]: in setup_suite get account_id for committee members from global_properties

File: suites/Operations/ProposalOperations/ProposalDelete.py [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+        operation = self.echo_ops.get_proposal_create_operation(
+            echo=self.echo,
+            fee_paying_account="1.2.6",
+            proposed_ops=collected_operation,
+            expiration_time=self.get_expiration_time(1000),
+            review_period_seconds=10,
+            signer=INIT0_PK
+        )
+        broadcast_result = self.echo_ops.broadcast(echo=self.echo, list_operations=operation)
+        proposal_id = broadcast_result["trx"]["operation_results"][0][1]
```
Comment (yauheni-vasilyeu) [inline suites/Operations/ProposalOperations/ProposalDelete.py:68]: 73-78 change for **operation is completed**

File: suites/Operations/ProposalOperations/ProposalUpdate.py [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+
+from common.base_test import BaseTest
+
+SUITE = {
+    "description": "Operation 'proposal_update'"
+}
+
+
+@lcc.prop("main", "type")
+@lcc.tags("operations", "asset_transfer_operations", "proposal_update")
```
Comment (yauheni-vasilyeu) [inline suites/Operations/ProposalOperations/ProposalUpdate.py]: asset_transfer_operations => change

File: suites/Operations/ProposalOperations/ProposalUpdate.py [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+class ProposalUpdate(BaseTest):
+
+    def __init__(self):
+        super().__init__()
+        self.__database_api_identifier = None
+        self.__registration_api_identifier = None
+        self.echo_acc0 = None
+        self.echo_acc1 = None
+
+    def get_expiration_time(self, seconds):
```
Comment (yauheni-vasilyeu) [inline suites/Operations/ProposalOperations/ProposalUpdate.py]: like in proposal_create suite

File: suites/Operations/ProposalOperations/ProposalUpdate.py [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+    def setup_suite(self):
+        super().setup_suite()
+        self._connect_to_echopy_lib()
+        lcc.set_step("Setup for {}".format(self.__class__.__name__))
+        self.__database_api_identifier = self.get_identifier("database")
+        self.__registration_api_identifier = self.get_identifier("registration")
+        lcc.log_info(
+            "API identifiers are: database='{}', registration='{}'".format(self.__database_api_identifier,
+                                                                           self.__registration_api_identifier))
+        self.echo_acc0 = self.get_account_id(self.accounts[0], self.__database_api_identifier,
```
Comment (yauheni-vasilyeu) [inline suites/Operations/ProposalOperations/ProposalUpdate.py]: remove default accounts and init committee_accounts (like in proposal_delete suite)

File: suites/Operations/ProposalOperations/ProposalUpdate.py [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+        )
+        broadcast_result = self.echo_ops.broadcast(echo=self.echo, list_operations=operation)
+        proposal_id = broadcast_result["trx"]["operation_results"][0][1]
+        lcc.log_info("Propsal created, proposal id: '{}'".format(proposal_id))
+
+        lcc.set_step("Get proposal before update")
+        response_id = self.send_request(self.get_request("get_proposed_transactions", ["1.2.7"]),
+                                        self.__database_api_identifier)
+        result = self.get_response(response_id)["result"][0]
+        check_that("required_active_approvals", result["required_active_approvals"], equal_to(["1.2.7"]))
```
Comment (yauheni-vasilyeu) [inline suites/Operations/ProposalOperations/ProposalUpdate.py]: change for operation_is_completed

File: suites/Operations/ProposalOperations/ProposalUpdate.py [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+            fee_paying_account="1.2.7",
+            proposal=proposal_id,
+            active_approvals_to_add=["1.2.7"],
+            active_approvals_to_remove=[],
+            key_approvals_to_add=[],
+            key_approvals_to_remove=[],
+            signer=[INIT1_PK]
+        )
+        collected_operation = self.collect_operations(operation, self.__database_api_identifier)
+        broadcast_result = self.echo_ops.broadcast(echo=self.echo, list_operations=collected_operation)
```
Comment (yauheni-vasilyeu) [inline suites/Operations/ProposalOperations/ProposalUpdate.py:83]: add operation_is_completed

**Review by yauheni-vasilyeu — CHANGES_REQUESTED** [review]

**Review by yauheni-vasilyeu — APPROVED** [review]

**Review by dbulka — APPROVED** [review]

**Pull request merged**